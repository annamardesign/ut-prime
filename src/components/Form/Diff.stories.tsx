import React from 'react';
import type { Story, Meta } from '@storybook/react';

import page from './README.mdx';
import type { Props } from './Form.types';
import Form from './index';
import decorators from '../test/decorator';
import {previous, current} from '../test/diff';
import {middleware} from '../Text/Text.mock';
import useForm from '../hooks/useForm';
import useLayout from '../hooks/useLayout';

const meta: Meta = {
    title: 'Form',
    component: Form,
    parameters: {docs: {page}},
    decorators,
    args: {
        state: {},
        middleware: [middleware]
    }
};
export default meta;

declare type StoryTemplate = Story<Partial<Props>> & {
    play: (context: {canvasElement: HTMLElement}) => Promise<void>
}

const Template: Story<Props> = args => {
    const formApi = useForm();
    return <div className='flex' style={{overflowX: 'hidden', width: '100%'}}>
        <Form formApi={formApi} layoutState={useLayout(args.schema, args.cards, args.layout, args.editors, undefined, args.layoutFields)} {...args} />
    </div>;
};

export const Diff: StoryTemplate = Template.bind({});

Diff.args = {
    schema: {
        properties: {
            previous: {
                title: '',
                widget: {
                    type: 'json'
                }
            },
            current: {
                title: '',
                widget: {
                    type: 'json'
                }
            }
        }
    },
    cards: {
        previous: {
            className: 'md:col-3',
            label: 'Previous',
            widgets: ['previous']
        },
        current: {
            className: 'md:col-3',
            label: 'Current',
            widgets: ['current']
        },
        diff: {
            className: 'md:col-3',
            label: 'Diff',
            widgets: [{
                name: 'current',
                parent: 'previous'
            }]
        }
    },
    layout: ['previous', 'current', 'diff'],
    value: {
        previous,
        current
    },
    onSubmit: () => {}
};
export const DiffBG: StoryTemplate = Template.bind({});
DiffBG.args = {
    ...Diff.args,
    lang: 'bg'
};

export const DiffAR: StoryTemplate = Template.bind({});
DiffAR.args = {
    ...Diff.args,
    lang: 'ar',
    dir: 'rtl'
};
