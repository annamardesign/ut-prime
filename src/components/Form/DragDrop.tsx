import React, { ReactElement } from 'react';
import { useDrag, useDrop } from 'react-dnd';
import type { DropTargetMonitor } from 'react-dnd';
import { Card } from '../prime';
import clsx from 'clsx';

const FIELD = Symbol('field');
type FieldMonitor = DropTargetMonitor<{
    card: string,
    label: string,
    index: number
}>;
const CARD = Symbol('card');
type CardMonitor = DropTargetMonitor<{
    card: string,
    title: string,
    index: [number, number | false]
}>;

export function DDField({design, children = null, name, index, card = '', move = null, labelClass = 'p-col-12 p-md-4', label, ...props}) {
    let labelProps;
    let isDragging = false;
    if (design) {
        if (card) {
            const [collected, dragField] = useDrag(
                () => ({
                    type: FIELD,
                    item: { index, card, label },
                    collect: monitor => ({
                        isDragging: !!monitor.isDragging()
                    })
                }),
                [card, index]
            );
            isDragging = collected.isDragging;
            labelProps = {
                ref: dragField,
                style: {cursor: 'move'}
            };
        }
        const [{ canDrop, isOver }, dropField] = useDrop(
            () => ({
                accept: FIELD,
                canDrop: () => !isDragging && move,
                drop: item => move('field', item, {index, card}),
                collect: monitor => ({
                    isOver: !!monitor.isOver(),
                    canDrop: !!monitor.canDrop()
                })
            }),
            [card, index, isDragging, move]
        );
        props.ref = dropField;
        props.style = {
            ...isDragging && {opacity: 0.5},
            ...!isDragging && canDrop && {
                ...isOver && {background: '#00ffff80'},
                outline: '1px dashed #00ffff80'
            }
        };
    }
    return (
        <div {...props}>
            {
                label ? <label className={labelClass} htmlFor={name} {...labelProps}>
                    {label}
                </label> : null
            }
            {children}
        </div>
    );
}

export function DDCard({children = null, card, index, move = undefined, flex = '', design, drop = false, title, ...props}) {
    if (drop && !design) return null;
    let dropZone: ReactElement<HTMLDivElement>;
    let divProps;
    if (design) {
        const [{isDragging}, dragCard] = useDrag(
            () => ({
                type: CARD,
                item: { card, index, title },
                collect: monitor => ({
                    isDragging: !!monitor.isDragging()
                })
            }),
            [card, index[0], index[1]]
        );
        const [{ isOverField, canDropField, dropLabel }, dropField] = useDrop(
            () => ({
                accept: FIELD,
                canDrop: (item: {card: string}) => item.card !== card,
                drop: item => move('field', item, {index: 1000, card}),
                collect: (monitor: FieldMonitor) => ({
                    isOverField: !!monitor.isOver(),
                    canDropField: !!monitor.canDrop(),
                    dropLabel: monitor.getItem()?.label
                })
            }),
            [card]
        );
        const [{ isOverCard, canDropCard, dragTitle }, dropCard] = useDrop(
            () => ({
                accept: CARD,
                canDrop: (item: {index: [any, any]}) => move && (item.index[0] !== index[0] || (!drop && item.index[1] !== index[1])),
                drop: item => move('card', item, {card, index}),
                collect: (monitor: CardMonitor) => ({
                    isOverCard: !!monitor.isOver(),
                    canDropCard: !!monitor.canDrop(),
                    dragTitle: monitor.getItem()?.title
                })
            }),
            [card, index[0], index[1], drop, move]
        );
        props.style = {
            ...isDragging && {opacity: 0.5},
            ...!isDragging && canDropCard && {
                ...isOverCard && {background: '#00ffff80'},
                outline: '1px dashed #00ffff80'
            },
            ...drop && {opacity: 0.5},
            ...drop && !canDropCard && {
                display: 'none'
            }
        };
        divProps = {
            ref: dropCard
        };
        title = (drop && canDropCard) ? 'ðŸ‘‰ ' + dragTitle : <div ref={dragCard} style={{cursor: 'move'}}>{title}</div>;
        dropZone = canDropField && <div
            ref={dropField}
            className={clsx('p-field p-grid', flex)}
            style={{
                background: isOverField ? '#00ffff80' : 'transparent',
                outline: '1px dashed #00ffff80'
            }}
        >
            <label className='p-col-12'>ðŸ‘‰ {dropLabel}</label>
        </div>;
    }
    return (
        <div {...divProps}>
            <Card title={title} {...props}>
                {children}
                {dropZone}
            </Card>
        </div>
    );
}
