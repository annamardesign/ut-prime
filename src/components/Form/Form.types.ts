import {WithStyles, withStyles, createStyles} from '@material-ui/core/styles';
import React from 'react';
import type { Schema as Validation } from 'joi';

import {Schema, Editors, Cards, Dropdowns} from '../types';
export interface Props extends Omit<React.HTMLAttributes<HTMLFormElement>, 'onSubmit'> {
    className?: string;
    schema: Schema;
    editors?: Editors;
    cards: Cards;
    dropdowns?: Dropdowns,
    layout?: (string | string[])[];
    loading?: string;
    validation?: Validation;
    design?: boolean;
    onSubmit: (data: {form: {}, submit: {}}) => void;
    setTrigger?: (trigger: (event: {}) => void) => void;
    value?: any;
}

const styles = createStyles({
    component: {
        '& .p-datatable-wrapper': {
            overflowX: 'auto'
        }
    }
});

export const Styled = withStyles(styles);
export type StyledType = React.FC<Props & WithStyles<typeof styles>>
