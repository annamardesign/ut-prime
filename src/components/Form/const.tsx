export const CHANGE = Symbol('change');
export const INDEX = Symbol('index');
