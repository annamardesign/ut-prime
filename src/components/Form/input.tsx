import React from 'react';
import { InputText, Password, InputTextarea, Dropdown, MultiSelect, TreeSelect, TreeTable, InputMask, InputNumber, Calendar, Checkbox, Skeleton, Column } from '../prime';
import { RefCallBack } from 'react-hook-form';

import Table from './inputs/Table';
import MultiSelectPanel from './inputs/MultiSelectPanel';
import SelectButton from './inputs/SelectButton';

export default function input(
    field: {
        onChange: (...event: any[]) => void;
        onBlur: () => void;
        value: any;
        name: string;
        ref: RefCallBack;
        className: string;
    }, {
        type = 'string',
        dropdown = '',
        parent = '',
        optionsFilter = null,
        title = '',
        ...props
    }: any = {id: field?.name},
    schema,
    dropdowns,
    filter,
    loading: string
) {
    if (loading) return <Skeleton className='p-inputtext'/>;
    const filterBy = item => (!filter && !optionsFilter) || Object.entries({...optionsFilter, ...filter}).every(([name, value]) => String(item[name]) === String(value));
    switch (type) {
        case 'dropdownTree': return <TreeSelect
            {...field}
            {...props}
        />;
        case 'text': return <InputTextarea
            {...field}
            {...props}
        />;
        case 'mask': return <InputMask
            {...field}
            {...props}
        />;
        case 'currency': return <InputNumber
            {...field}
            inputRef={field.ref}
            onChange={e => {
                field.onChange?.(e.value);
            }}
            minFractionDigits={2}
            maxFractionDigits={4}
            {...props}
        />;
        case 'boolean': return <Checkbox
            {...field}
            inputRef={field.ref}
            onChange={e => field.onChange?.(e.checked)}
            checked={field.value}
            {...props}
        />;
        case 'table': return <Table
            {...field}
            selectionMode='checkbox'
            filter={filter}
            properties={schema?.items?.properties}
            dropdowns={dropdowns}
            {...props}
        />;
        case 'dropdown': return <Dropdown
            {...field}
            options={dropdowns?.[dropdown]?.filter(filterBy) || []}
            disabled={parent && !filter}
            {...props}
        />;
        case 'multiSelect': return <MultiSelect
            {...field}
            options={dropdowns?.[dropdown]?.filter(filterBy) || []}
            disabled={parent && !filter}
            display='chip'
            {...props}
        />;
        case 'select': return <SelectButton
            {...field}
            options={dropdowns?.[dropdown]?.filter(filterBy) || []}
            disabled={parent && !filter}
            {...props}
        />;
        case 'multiSelectTree': return <TreeSelect
            {...field}
            options={dropdowns?.[dropdown]?.filter(filterBy) || []}
            disabled={parent && !filter}
            display='chip'
            selectionMode='multiple'
            metaKeySelection={false}
            onChange={e => {
                field.onChange?.(Object.keys(e.value));
            }}
            value={field.value?.map && Object.fromEntries(field.value?.map(value => [value, true]))}
            {...props}
        />;
        case 'multiSelectPanel': return <MultiSelectPanel
            appendTo='self'
            {...field}
            options={dropdowns?.[dropdown]?.filter(filterBy) || []}
            disabled={parent && !filter}
            {...props}
        />;
        case 'selectTable': {
            const all = dropdowns?.[dropdown];
            const value = all.filter(filterBy) || [];
            const dataKey = props.dataKey || 'value';
            const valueKeys = value.map(item => item[dataKey]);
            const single = props.selectionMode === 'single';
            const hidden = !single && (field.value || []).filter(item => !valueKeys.includes(item));
            return <Table
                {...field}
                actions={{allowAdd: false, allowEdit: false, allowDelete: false}}
                dataKey={dataKey}
                value={value}
                selection={single ? value.find(row => row?.[dataKey] === field.value) : value.filter(row => field.value?.includes(row?.[dataKey]))}
                onSelectionChange={event =>
                    field.onChange?.(
                        single
                            ? event.value[dataKey]
                            : [].concat(hidden, event.value?.map(row => row?.[dataKey])),
                        {children: false, ...props.change}
                    )
                }
                {...props}
            >
                <Column field='label' header={title}/>
            </Table>;
        }
        case 'multiSelectTreeTable': return <TreeTable
            {...field}
            value={dropdowns?.[dropdown]?.filter(filterBy) || []}
            selectionKeys={field.value}
            onSelectionChange={e => field.onChange?.(e.value)}
            selectionMode='checkbox'
            {...props}
        >
            <Column field='label' expander/>
        </TreeTable>;
        case 'date-time': return <Calendar
            {...field}
            showTime
            showIcon
            value={field.value != null ? new Date(field.value) : field.value}
            {...props}
        />;
        case 'time': return <Calendar
            {...field}
            timeOnly
            showIcon
            value={field.value != null ? new Date(field.value) : field.value}
            {...props}
        />;
        case 'date': return <Calendar
            {...field}
            showIcon
            value={field.value != null ? new Date(field.value) : field.value}
            {...props}
        />;
        case 'integer': return <InputNumber
            {...field}
            inputClassName='w-full'
            inputRef={field.ref}
            onChange={e => {
                field.onChange?.(e.value);
            }}
            showButtons
            {...props}
        />;
        case 'password': return <Password
            {...field}
            {...props}
        />;
        default: return <InputText
            {...field}
            {...props}
        />;
    }
}
