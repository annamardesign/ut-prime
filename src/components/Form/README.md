# Form

## How to use

```jsx
import Form from 'ut-prime/core/Form';

<Form className='name' />;
```

## Props

- **className** - class applied to root `Form` html element.

| propName  | propType | defaultValue | isRequired |
| --------- | -------- | ------------ | ---------- |
| className | string   |              | no         |
