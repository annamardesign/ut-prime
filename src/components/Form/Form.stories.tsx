import React from 'react';
import type { Story, Meta } from '@storybook/react';
import { userEvent } from '@storybook/testing-library';
import { within } from '@testing-library/react';

// @ts-ignore: md file and not a module
import page from './README.mdx';
import type { Props } from './Form.types';
import Form from './index';
import tree from '../test/tree';
import {input, dropdowns} from '../test/input';

const meta: Meta = {
    title: 'Form',
    component: Form,
    parameters: {docs: {page}},
    args: {
        state: {}
    }
};
export default meta;

const Template: Story<Props> = args =>
    <div className='flex' style={{overflowX: 'hidden', width: '100%'}}>
        <Form {...args} />
    </div>;

export const Basic: Story<Props> = Template.bind({});

Basic.args = {
    ...tree,
    layout: ['edit', ['taxonomy', 'reproduction'], 'links', 'morphology'],
    dropdowns: {'tree.type': [{value: 1, label: 'Conifer'}, {value: 2, label: 'Broadleaf'}]},
    value: {tree: {treeName: 'Oak', treeId: 1, treeType: 1}},
    onSubmit: () => {}
};

export const Input: Story<Props> = Template.bind({});

Input.args = {
    ...input,
    layout: ['left', 'center', 'right'],
    dropdowns: dropdowns,
    value: {input: {}},
    onSubmit: () => {}
};

Input.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    const body = within(document.body);
    const type = (role, id, text) => userEvent.type(canvas.getByRole(role, {name: (name, el) => el.id === id}), text);
    const click = (id) => userEvent.click(canvas.getByTestId(id));
    const clickOption = (id, name, role = 'option') => {
        id && click(id);
        body.getByRole(role, {name}).click();
    };
    const clickWithin = (id, name, role = 'option') => within(canvas.getByTestId(id)).getByRole(role, {name}).click();

    // left
    type('textbox', 'input-input', 'input');
    type('textbox', 'input-text', 'text');
    type('textbox', 'input-mask', '192168000001');
    type('textbox', 'input-date', '01/31/2022');
    type('textbox', 'input-time', '20:00');
    click('input-boolean');
    type('textbox', 'input-datetime', '01/31/2022 20:00');
    type('spinbutton', 'input-currency', '1234567.89');
    type('spinbutton', 'input-number', '12345.67890');
    type('spinbutton', 'input-integer', '1234567890');
    type('textbox', 'input-password', '123');

    // center
    clickOption('input-dropdown', 'EUR');
    clickOption('input-dropdownTree', 'Asia', 'treeitem');
    clickOption('input-multiSelect', 'Rome');
    click('input-multiSelect'); // close the multiselect dropdown
    clickOption('input-multiSelectTree', 'Earth', 'treeitem');
    click('input-multiSelectTree'); // close the multiselect dropdown
    click('input-table-addButton');
    type('textbox', 'input-table-0-name', 'name');
    type('textbox', 'input-table-0-value', 'value');

    // right
    clickWithin('input-select', 'One', 'button');
    clickWithin('input-selectTable', 'One', 'cell');
    clickWithin('input-multiSelectPanel', 'One', 'option');
    within(within(canvas.getByTestId('input-multiSelectTreeTable')).getByRole('row', {name: 'One'})).getAllByRole('checkbox')[0].click();
};

export const Table: Story<Props> = Template.bind({});
Table.args = {
    ...input,
    layout: ['table'],
    dropdowns: dropdowns,
    value: {table: [{input: 'test'}]},
    onSubmit: () => {}
};

Table.play = async({canvasElement}) => {
    const canvas = within(canvasElement);
    (await within(within(canvas.getByTestId('table1')).getByRole('table')).findByRole('button')).click();
    within(canvas.getByTestId('table2')).getByRole('button', {name: ''}).click();
    within(canvas.getByTestId('table3')).getByRole('button', {name: ''}).click();
    within(canvas.getByTestId('table4')).getByRole('button', {name: ''}).click();
};
