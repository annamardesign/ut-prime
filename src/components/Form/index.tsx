import React from 'react';
import clsx from 'clsx';
import Joi from 'joi';
import get from 'lodash.get';
import { joiResolver } from '@hookform/resolvers/joi';

import { Card } from '../prime';
import { Styled, StyledType } from './Form.types';
import { Properties } from '../types';
import useForm from '../hooks/useForm';
import Controller from '../Controller';
import input from './input';

const getSchema = (properties: Properties) : Joi.Schema => Object.entries(properties).reduce(
    (schema, [name, field]) => {
        if ('properties' in field) {
            return schema.append({[name]: getSchema(field.properties)});
        } else {
            const {title, validation = Joi.string().min(0).allow('', null)} = field;
            return schema.append({[name]: validation.label(title || name)});
        }
    },
    Joi.object()
);

const getIndex = (properties: Properties, root: string) : Properties => Object.entries(properties).reduce(
    (map, [name, property]) => {
        return ('properties' in property) ? {
            ...map,
            ...getIndex(property.properties, name + '.')
        } : {
            ...map,
            [root + name]: property
        };
    },
    {}
);

interface Errors {
    message?: string
}

const flat = (e: Errors, path = '') => Object.entries(e).map(
    ([name, value]) => typeof value.message === 'string' ? (path ? path + '.' + name : name) : flat(value, name)
);

const Form: StyledType = ({ classes, className, properties, cards, layout, onSubmit, setTrigger, value, dropdowns, validation, ...rest }) => {
    const joiSchema = validation || getSchema(properties);
    const index = getIndex(properties, '');
    const {handleSubmit, control, reset, formState: {errors, isDirty}, watch, setValue} = useForm({resolver: joiResolver(joiSchema)});
    const getFormErrorMessage = (name) => {
        const error = get(errors, name);
        return error && <small className="p-error">{error.message}</small>;
    };
    React.useEffect(() => {
        if (setTrigger) setTrigger(isDirty ? prev => handleSubmit(onSubmit) : undefined);
    }, [setTrigger, handleSubmit, onSubmit, isDirty]);
    React.useEffect(() => {
        reset(value || {});
    }, [value]);

    const children: {[parent: string]: string[]} = {};

    function addParent(properties: Properties, name: string) {
        const parent = properties[name]?.editor?.parent;
        if (parent) {
            const items = children[parent];
            if (items) items.push(name);
            else children[parent] = [name];
            return watch(parent);
        }
    }

    const parentChange = (name : string) => {
        const items = children[name];
        if (items) items.forEach(child => setValue(child, null));
    };

    function card(id: string) {
        const {title, properties = [], flex} = (cards[id] || {title: '❌ ' + id});
        return (
            <Card title={title} key={id} className='p-fluid p-mb-3'>
                <div className={clsx(flex && 'p-d-flex p-flex-wrap')}>
                    {properties.map(name => index[name]
                        ? <div className={clsx('p-field p-grid', flex)} key={name}>
                            {index[name].title ? <label className='p-col-12 p-md-4'>
                                {index[name].title}
                            </label> : null}
                            <div className={clsx(index[name].title ? 'p-col-12 p-md-8' : 'p-col-12')}>
                                <Controller
                                    control={control}
                                    name={name}
                                    render={
                                        ({field}) => input(
                                            {
                                                className: clsx({ 'p-invalid': errors[name] }),
                                                ...field,
                                                onChange: e => {
                                                    try {
                                                        parentChange(name);
                                                    } finally {
                                                        field.onChange(e);
                                                    }
                                                }
                                            },
                                            index[name].editor,
                                            index[name],
                                            dropdowns,
                                            addParent(index, name)
                                        )
                                    }
                                />
                            </div>
                            {getFormErrorMessage(name)}
                        </div>
                        : <div className="p-field p-grid" key={name}>❌ {name}</div>
                    )}
                </div>
            </Card>
        );
    }
    const visibleCards = (layout || Object.keys(cards));
    const errorFields = flat(errors).flat();
    const visibleProperties = visibleCards.map(id => {
        const nested = [].concat(id);
        return nested.map(cardName => cards[cardName]?.properties);
    }).flat(10).filter(Boolean);

    return (
        <form {...rest} onSubmit={handleSubmit(onSubmit)} className={clsx('p-grid p-col p-mt-2', className)}>
            {
                !!Object.keys(errors).length && <div className='p-col-12'>
                    {errorFields.map(name => !visibleProperties.includes(name) && <><small className="p-error">{get(errors, name)?.message}</small><br /></>)}
                </div>
            }
            {visibleCards.map(id => {
                const nested = [].concat(id);
                const key = nested[0];
                return (
                    <div key={key} className={clsx('p-col-12', cards[key]?.className || 'p-xl-6')}>
                        {nested.map(card)}
                    </div>
                );
            })}
        </form>
    );
};

export default Styled(Form);
