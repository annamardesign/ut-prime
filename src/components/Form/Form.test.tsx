import React from 'react';

import { render } from '../test';
import { Basic, Input } from './Form.stories';

describe('<Form />', () => {
    it('render equals snapshot', async() => {
        const { findByTestId } = render(<Basic />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
    it('render equals snapshot', async() => {
        const { findByTestId } = render(<Input />);
        expect(await findByTestId('ut-front-test')).toMatchSnapshot();
    });
});
