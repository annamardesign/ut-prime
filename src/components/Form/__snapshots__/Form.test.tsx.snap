// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Form /> render equals snapshot 1`] = `
<div
  data-testid="ut-front-test"
>
  <div
    class="flex"
    style="overflow-x: hidden; width: 100%;"
  >
    <div
      class="p-toast p-component p-toast-top-right"
    >
      <div />
    </div>
    <form
      class="grid col align-self-start Form-component-1"
    >
      <div
        class="col-12 xl:col-6"
      >
        <div>
          <div
            class="p-card p-component card mb-3"
            id="edit"
          >
            <div
              class="p-card-body"
            >
              <div
                class="p-card-title"
              >
                Tree
              </div>
              <div
                class="p-card-content"
              >
                <div
                  class=""
                >
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.treeName"
                    >
                      Name
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <input
                        class="p-inputtext p-component p-filled w-full"
                        id="tree.treeName"
                        name="tree.treeName"
                        value="Oak"
                      />
                    </div>
                  </div>
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.treeDescription"
                    >
                      Description
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <textarea
                        class="p-inputtextarea p-inputtext p-component w-full"
                        id="tree.treeDescription"
                        name="tree.treeDescription"
                      />
                    </div>
                  </div>
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.treeType"
                    >
                      Type
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <div
                        class="p-dropdown p-component p-inputwrapper w-full p-inputwrapper-filled"
                        id="tree.treeType"
                      >
                        <div
                          class="p-hidden-accessible"
                        >
                          <input
                            aria-haspopup="listbox"
                            readonly=""
                            type="text"
                          />
                        </div>
                        <div
                          class="p-hidden-accessible p-dropdown-hidden-select"
                        >
                          <select
                            aria-hidden="true"
                            name="tree.treeType"
                            tabindex="-1"
                          >
                            <option
                              value=""
                            />
                            <option
                              value="1"
                            >
                              Conifer
                            </option>
                          </select>
                        </div>
                        <span
                          class="p-dropdown-label p-inputtext"
                        >
                          Conifer
                        </span>
                        <div
                          aria-expanded="false"
                          aria-haspopup="listbox"
                          class="p-dropdown-trigger"
                          role="button"
                        >
                          <span
                            class="p-dropdown-trigger-icon p-clickable pi pi-chevron-down"
                          />
                        </div>
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-12 xl:col-6"
      >
        <div>
          <div
            class="p-card p-component card mb-3"
            id="taxonomy"
          >
            <div
              class="p-card-body"
            >
              <div
                class="p-card-title"
              >
                Taxonomy
              </div>
              <div
                class="p-card-content"
              >
                <div
                  class=""
                />
              </div>
            </div>
          </div>
        </div>
        <div>
          <div
            class="p-card p-component card mb-3"
            id="reproduction"
          >
            <div
              class="p-card-body"
            >
              <div
                class="p-card-title"
              >
                Reproduction
              </div>
              <div
                class="p-card-content"
              >
                <div
                  class=""
                >
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.seedDescription"
                    >
                      Seed
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <input
                        class="p-inputtext p-component w-full"
                        id="tree.seedDescription"
                        name="tree.seedDescription"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.flowerDescription"
                    >
                      Flower
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <input
                        class="p-inputtext p-component w-full"
                        id="tree.flowerDescription"
                        name="tree.flowerDescription"
                        value=""
                      />
                    </div>
                  </div>
                  <div
                    class="field grid"
                  >
                    <label
                      class="col-12 md:col-4"
                      for="tree.fruitName"
                    >
                      Fruit
                    </label>
                    <div
                      class="col-12 md:col-8"
                    >
                      <input
                        class="p-inputtext p-component w-full"
                        id="tree.fruitName"
                        name="tree.fruitName"
                        value=""
                      />
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="col-12 xl:col-6"
      >
        <div>
          <div
            class="p-card p-component card mb-3"
            id="morphology"
          >
            <div
              class="p-card-body"
            >
              <div
                class="p-card-title"
              >
                Morphology
              </div>
              <div
                class="p-card-content"
              >
                <div
                  class=""
                />
              </div>
            </div>
          </div>
        </div>
      </div>
    </form>
  </div>
</div>
`;
